cli/cmd/identity.go

```diff
@@ -0,0 +1,295 @@
+		fetches the tls certificate.
+		`,
+		Example: ` 
+		#Get certificate from pod foo-bar in the default namespace.
+		linkerd identity pod/foo-bar
+
+		#Get certificates from all pods in emojivoto namespace
+		linkerd identity -n emojivoto
+		
+		#Get certificate from all pods with name=nginx
+		linkerd identity -l name=nginx
+		`,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			k8sAPI, err := k8s.NewAPI(kubeconfigPath, kubeContext, impersonate, impersonateGroup, 0)
+			if err != nil {
+				return err
+			}
+
+			resources, err := resource.FetchKubernetesResources(cmd.Context(), k8sAPI, metav1.ListOptions{LabelSelector: k8s.ControllerNSLabel})
+			if err != nil {
+				return err
+			}
+			// check if linkerd control plane exists
+			if len(resources) == 0 {
+				return fmt.Errorf("Cannot find Linkerd\nValidate the install with: linkerd check")
+			}
+
+			pods, err := getPods(cmd.Context(), k8sAPI, options.namespace, options.selector, args)
+			if err != nil {
+				return err
+			}
+
+			results := getCertificate(k8sAPI, pods, k8s.ProxyAdminPortName, 30*time.Second, emitLog)
+
+			for i, result := range results {
+				fmt.Printf("\nPOD %s (%d of %d)\n\n", result.pod, i+1, len(results))
+				if result.err == nil {
+					fmt.Println("Version: ", result.Certificate[0].Version)
+					fmt.Println("Serial Number: ", result.Certificate[0].SerialNumber)
+					fmt.Println("Signature Algorithm: ", result.Certificate[0].SignatureAlgorithm)
+					fmt.Println("Validity")
+					fmt.Println("\t Not Before: ", result.Certificate[0].NotBefore.Format(time.RFC3339))
+					fmt.Println("\t Not After: ", result.Certificate[0].NotAfter.Format(time.RFC3339))
+					fmt.Println("Subject: ", result.Certificate[0].Subject)
+					fmt.Println("Issuer:  ", result.Certificate[0].Issuer)
+					fmt.Println("Subject Public Key Info:")
+					fmt.Println("\tPublic Key Algorithm: ", result.Certificate[0].PublicKeyAlgorithm)
+					fmt.Println("Signature: \n", result.Certificate[0].Signature)
+					resultCerticate := tls1.EncodeCertificatesPEM(result.Certificate[0])
+
+					fmt.Printf("\n%s", resultCerticate)
+				} else {
+					fmt.Printf("\n%s\n", result.err)
+				}
+			}
+			return nil
+		},
+	}
+
+	cmd.PersistentFlags().StringVarP(&options.namespace, "namespace", "n", options.namespace, "Namespace to use for --proxy versions (default: all namespaces)")
+	cmd.PersistentFlags().StringVarP(&options.selector, "selector", "l", options.selector, "Selector (label query) to filter on, supports ‘=’, ‘==’, and ‘!=’ ")
+	return cmd
+}
+
+// getCertificate fetches the certificates for all the pods
+func getCertificate(k8sAPI *k8s.KubernetesAPI, pods []corev1.Pod, portName string, waitingTime time.Duration, emitLog bool) []certificate {
+	var certificates []certificate
+	resultChan := make(chan certificate)
+	var activeRoutines int32
+	for _, pod := range pods {
+		atomic.AddInt32(&activeRoutines, 1)
+		go func(p corev1.Pod) {
```

Comment:
I think it's fine to fetch each certificate in series rather than doing them in parallel.  This should let us simplify this and get rid of the channel.  It also happens to fix the bug where if there are no pods selected by the label selector, the select hangs.